Add Gamma distribution functions

talgalili opened the pull request:
For importance, see: https://en.wikipedia.org/wiki/Gamma_distribution.

This diff aims to revive this (2 years old) issue: #15961. (this is a fixed version of this diff: #15960)

Test plan: added tests.


```
== RELEASE NOTES ==

General Changes
* Add gamma_cdf() and inverse_gamma_cdf() functions.
```

talgalili commented:
So good news - all the required tests passed. :)

talgalili commented:
I'm trying to understand the two failures.
1. https://github.com/prestodb/presto/actions/runs/4418357600/jobs/7745370462
2. https://github.com/prestodb/presto/actions/runs/4418357592/jobs/7745376737

In both cases I can't see how the issue is related to this diff.

rschlussel commented:
> I'm trying to understand the two failures.
> 
> 1. https://github.com/prestodb/presto/actions/runs/4418357600/jobs/7745370462
> 2. https://github.com/prestodb/presto/actions/runs/4418357592/jobs/7745376737
> 
> In both cases I can't see how the issue is related to this diff.

Those failures just look like flaky tests

feilong-liu commented on presto-main/src/main/java/com/facebook/presto/operator/scalar/MathFunctions.java:
```diff
@@ -901,6 +902,36 @@ public static double chiSquaredCdf(
+        checkCondition(p >= 0 && p <= 1, INVALID_FUNCTION_ARGUMENT, "p must be in the interval [0, 1]");
+        checkCondition(shape > 0, INVALID_FUNCTION_ARGUMENT, "shape must be greater than 0");
+        checkCondition(scale > 0, INVALID_FUNCTION_ARGUMENT, "scale must be greater than 0");
+        GammaDistribution distribution = new GammaDistribution(null, shape, scale, GammaDistribution.DEFAULT_INVERSE_ABSOLUTE_ACCURACY);
```
Will there be any need of changing the accuracy when using this function? If yes, can make this a function argument too.
A total noob on math here, I see that the default accuracy is 1e-9, which I guess should be enough. Just a reminder if as a user you foresee such case.

talgalili replied:
Thanks @feilong-liu 

I doubt such a need would arise.
If it does, then it might be worth adding that argument to all the probability functions (there are a bunch).
I'm going to leave it as is for now.

feilong-liu commented on presto-main/src/test/java/com/facebook/presto/operator/scalar/TestMathFunctions.java:
```diff
@@ -1467,6 +1467,39 @@ public void testChiSquaredCdf()
+
+    @Test
+    public void testGammaCdf()
+            throws Exception
```
`Exception 'java.lang.Exception' is never thrown in the method `, remove this part

talgalili replied:
Thanks, done.
(might be worth then to remove it from other probability functions then. This was never mentioned before. But if to remove it from other functions in the file, I propose to do it in another diff)

feilong-liu reviewed (approved):
Looks good. Just some nits.

talgalili commented:
> Looks good. Just some nits.

Thanks for the review @feilong-liu 
I've made changes as you proposed, squashed and updated the diff.

Could you please review and merge?

feilong-liu commented:
> > Looks good. Just some nits.
> 
> Thanks for the review @feilong-liu I've made changes as you proposed, squashed and updated the diff.
> 
> Could you please review and merge?

cc @rschlussel

rschlussel commented:
Can you just fix the commit message to follow our guidelines. e.g. `Add Gamma distribution functions`.  Also, add attribution for the person who made the original commit by adding a coauthor line to the commit body.  https://github.com/prestodb/presto/wiki/Review-and-Commit-guidelines#commit-formatting-and-pull-requests

talgalili commented:
> Can you just fix the commit message to follow our guidelines. e.g. `Add Gamma distribution functions`. Also, add attribution for the person who made the original commit by adding a coauthor line to the commit body. https://github.com/prestodb/presto/wiki/Review-and-Commit-guidelines#commit-formatting-and-pull-requests

Thanks for the input @rschlussel 

Done:

![image](https://user-images.githubusercontent.com/976006/225412413-dc87c30d-010b-4ce5-95f2-23d8715373b5.png)

rschlussel commented:
the co-author line should look like below. That way github will parse it and give "credit" properly.

`Co-authored-by: Foo Bar <foo@bar.com>`

talgalili commented:
> the co-author line should look like below. That way github will parse it and give "credit" properly.
> 
> `Co-authored-by: Foo Bar <foo@bar.com>`

@rschlussel 

I believe that github recognizes also the use of at (@) + handle name.

See the screenshot that shows that the hover works:

![image](https://user-images.githubusercontent.com/976006/225417948-9c422108-8547-4569-bbaa-d2854e768d4e.png)

Furthermore, I see that @Masha-Iris has never shared her email publicly on her profile, so I'm not sure she'd want us to expose it publicly through this diff commit.
https://github.com/Masha-Iris

WDYT?

rschlussel commented:
so that tags them, but you want them to be recognized as a co-author.  See e.g. these commits where you can see github says "feilong-liu and souravpal committed on...": https://github.com/prestodb/presto/pull/18630/commits.  That's the behavior we're looking for, and it doesn't look like the tagging you have doing that, so you'll need to use the exact syntax Ishared above.

jainxrohit reviewed (approved):
Looks good. Please add co-author as Rebecca suggested.

talgalili commented:
> so that tags them, but you want them to be recognized as a co-author. See e.g. these commits where you can see github says "feilong-liu and souravpal committed on...": https://github.com/prestodb/presto/pull/18630/commits. That's the behavior we're looking for, and it doesn't look like the tagging you have doing that, so you'll need to use the exact syntax Ishared above.

@rschlussel 
That helped thanks.

I got it to work now (and without exposing her email, but by using the syntax in the reference diff)

![image](https://user-images.githubusercontent.com/976006/225423294-e894279d-3778-4d88-b004-7a07dc1fa498.png)


Please LMK if there is anything else needed to land the diff.

talgalili commented:
> Looks good. Please add co-author as Rebecca suggested.

@jainxrohit

Done.

talgalili commented:
I see all checks passed, and that the diff was approved by 3 people.
What else is needed for a merge?
Is it done automatically, or is there any extra needed step?

Thanks!

jainxrohit commented:
> I see all checks passed, and that the diff was approved by 3 people. What else is needed for a merge? Is it done automatically, or is there any extra needed step?
> 
> Thanks!

I have merged on your behalf.

talgalili commented:
Thanks @jainxrohit 
Was ot something I could have done myself?

jainxrohit commented:
> Thanks @jainxrohit Was ot something I could have done myself?

I think so. It needs some permissions, I am not sure if you have those.

talgalili commented:
Thanks everyone for the help.

I took the opportunity to do a followup diff to also add the F distribution:
https://github.com/prestodb/presto/pull/19203

(I also found a bunch of errors in the doc, I'll fix them in another diff after that one)
